File: Go/Day3/1.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	re := regexp.MustCompile(`mul\([0-9]+,[0-9]+\)`)

	text, _ := os.ReadFile("./Input/3.txt")

	times := re.FindAll(text, -1)
	total := 0

	for _, byte_arr := range times {
		str := string(byte_arr)

		str = strings.Replace(str, "mul(", "", -1)
		str = strings.Replace(str, ")", "", -1)
		split := strings.Split(str, ",")
		left, _ := strconv.Atoi(split[0])
		right, _ := strconv.Atoi(split[1])
		total += (left * right)
	}

	fmt.Println(total)
}

File: Python/Day3/1.py

text = ""
start = 0
total = 0
with open("./Input/3.txt") as f:
    text = f.read()

index = text.index("mul(")

def find_mul_content(start, text):
    pos = start
    content = ""
    while pos < len(text):
        if not text[pos].isdigit() and text[pos] != ',':
            break
        elif text[pos] != ")":
            content += text[pos]
            pos+=1
            continue
        else:
            break
    
    return (content, pos)

def process_content(content):
    split = content.split(",")
    if len(split) != 2:
        return 0
    if not split[0].isnumeric() or not split[1].isnumeric():
        return 0
    return int(split[0]) * int(split[1])

while start < len(text):
    print(f"Starting at {start}")
    print(f"Total at {total}")
    try:
        if start + 4 >= len(text):
            break
        
        content, start = find_mul_content(index + 4, text)
        total += process_content(content)
        index = text.index("mul(", start - 1)
    except:
        break



print(total)


